libs/community/langchain_community/retrievers/sentence_window_retriever.py

```diff
@@ -0,0 +1,184 @@
+from typing import Dict, List
+
+import numpy as np
+from langchain_core.callbacks import CallbackManagerForRetrieverRun
+from langchain_core.documents import Document
+from langchain_core.pydantic_v1 import root_validator
+from langchain_core.retrievers import BaseRetriever
+from langchain_core.vectorstores import VectorStore
+
+# from langchain_pinecone import PineconeVectorStore
+from langchain_community.vectorstores import FAISS, Chroma
+from langchain_community.vectorstores.faiss import (
+    dependable_faiss_import,
+)
+
+
+class SentenceWindowRetriever(BaseRetriever):
+    """`Sentence Window` retriever.
+
+    Sentence Window Retriever dissociates retrieval from generation
+    by appending adjacent chunks to the retrieved chunk to provide
+    additional context to the generation model
+
+    Currently, it supports 2 backends:
+    FAISS, Chroma
+    """
+
+    store: VectorStore
+    """VectorStore containing the chunks of text. Only supports the 
+        above listed databases"""
+
+    k: int = 4
+    """Number of top results to return"""
+
+    window_size: int = 2
+    """Number of adjacent chunks on each side to be added """
+
+    @root_validator
+    def check_database_type(cls, values: Dict) -> Dict:
+        k = values.get("k", -1)
+        window_size = values.get("window_size", -1)
+
+        if k < 0:
+            raise ValueError("The value of k must be greater than 0")
+
+        if window_size < 0:
+            raise ValueError("The value of window_size must be greater than 0")
+
+        return values
+
+    def _get_relevant_documents(
+        self, query: str, *, run_manager: CallbackManagerForRetrieverRun
+    ) -> List[Document]:
+        """Sync implementations for retriever."""
+
+        if type(self.store) == Chroma:
```

Comment:
MAJOR: Are you able to implement a vectostore agnostic solution? If we can't make it agnostic, we will likely not add this technique to the code base.

My understanding was that this could be solved by using two abstractions:

1) retriever: base retriever
2) document store: an entity that given a document id can return the document content


Would composing or sub-classing `ParentRetriever` work?

Reply:
The issue with forcing a vectorstore agnostic implementation is that each vectorstore has its own unique way of retrieving specific texts (either based on their ID or the new 'chunk_id' metadata variable). 

One way we can make it mostly agnostic is by forcing users to add a specific ID variable to chunks ( as the database's ID and not the chunk_id in the metadata). So for the databases that support filtering based on ID, we can have a general solution. 

But for the databases, that dont support such filtering, we might have need a specific implementation. If its crucial to not have custom solutions, then we could decide not to support those databases for this retriever.

Reply:
We generally want the implementations to be generic otherwise users will not be able to use them, they become hard to maintain and are hard to test / debug. 

It sounds like before this can be implemented, we need to improve the vectorstore interface first. 

What functionality do you need from vectorstores?

My understanding is that either one of two techniques will work:

```python
AND = {
   'source_id': source_id,
   'chunk_id' : {
     'between': (chunk_id-window_size, chunk_id+window_size)
   }
}
```

OR just a function to get documents by ids: `get_documents_by_ids`

---

The second case requires populating the metadata of each document with information about its nearby neighbors (including distance and doc ids).

---

By the way, you can work around this right now by combining BaseStore and VectorStore abstractions.

Reply:
So similar to `embeddings.embed_query` which is agnostic of the vectorstore implementation, if there is a common vectorstore functionality that allows for filtering on chunk ID (either the database id or the chunk_id in metadata) wherever available. 

So the `get_documents_by_ids` would be something that could ideally achieve that. The first option depends on the filtering syntax for each database. Unless we implement our own filtering syntax that converts filter arguments to the database specific filtering syntax. That would allow us to have an agnostic implementation for this retriever. 

Im not sure how the BaseStore + VectorStore would help work around the database specific syntax requirements? If you could share more details on that I can try implementing that.

Reply:
`BaseStore` allows getting content by ID -- so it provides a way of doing `get_documents_by_id`

Reply:
Thanks. Let me take a look at that and see if I can implement `get_documents_by_id` for vectorstore.